### DISCUSSION TITLE
Where controls the actual directional ping width?

### DISCUSSION THREAD

**Discussion by iridescent-lightning** [FakeFishGames/Barotrauma#13174] [Code questions and modding support]
I was looking through the client-side sonar class. I couldn't find any code that is related to the width of the sonar. I planed to make the directional ping a bit wider. Devs could you provide some hints? Also, it seems that there is no real directional ping. You just don't see the rest of the part. Maybe I'm wrong on this. But how to only allow mainsub to blip when it is pinged? My current work around is to make the sub doesn't blip at all when enabled the directional ping, which is pretty neat as well.

**Comment by iridescent-lightning** [comment]
public static bool PingPrefix(Vector2 pingSource,  Vector2 transducerPos,  float pingRadius,  float prevPingRadius,  float displayScale,  float range,  bool passive,
             float pingStrength,  AITarget needsToBeInSector, Sonar __instance)
		{
			
			Sonar _ = __instance;
            float prevPingRadiusSqr = prevPingRadius * prevPingRadius;
            float pingRadiusSqr = pingRadius * pingRadius;
            
			
			
			//let us detect outside as well!
            /*inside a hull -> only show the edges of the hull
            if (_.item.CurrentHull != null && _.DetectSubmarineWalls)
            {
                _.CreateBlipsForLine(
                    new Vector2(_.item.CurrentHull.WorldRect.X, _.item.CurrentHull.WorldRect.Y), 
                    new Vector2(_.item.CurrentHull.WorldRect.Right, _.item.CurrentHull.WorldRect.Y), 
                    pingSource, transducerPos,
                    pingRadius, prevPingRadius, 50.0f, 5.0f, range, 2.0f, passive, needsToBeInSector: needsToBeInSector);

                _.CreateBlipsForLine(
                    new Vector2(_.item.CurrentHull.WorldRect.X, _.item.CurrentHull.WorldRect.Y - _.item.CurrentHull.Rect.Height),
                    new Vector2(_.item.CurrentHull.WorldRect.Right, _.item.CurrentHull.WorldRect.Y - _.item.CurrentHull.Rect.Height),
                    pingSource, transducerPos,
                    pingRadius, prevPingRadius, 50.0f, 5.0f, range, 2.0f, passive, needsToBeInSector: needsToBeInSector);

                _.CreateBlipsForLine(
                    new Vector2(_.item.CurrentHull.WorldRect.X, _.item.CurrentHull.WorldRect.Y),
                    new Vector2(_.item.CurrentHull.WorldRect.X, _.item.CurrentHull.WorldRect.Y - _.item.CurrentHull.Rect.Height),
                    pingSource, transducerPos,
                    pingRadius, prevPingRadius, 50.0f, 5.0f, range, 2.0f, passive, needsToBeInSector: needsToBeInSector);

                _.CreateBlipsForLine(
                    new Vector2(_.item.CurrentHull.WorldRect.Right, _.item.CurrentHull.WorldRect.Y),
                    new Vector2(_.item.CurrentHull.WorldRect.Right, _.item.CurrentHull.WorldRect.Y - _.item.CurrentHull.Rect.Height),
                    pingSource, transducerPos,
                    pingRadius, prevPingRadius, 50.0f, 5.0f, range, 2.0f, passive, needsToBeInSector: needsToBeInSector);

                return false;
            }*/
            foreach (Submarine submarine in Submarine.Loaded)
            {
				
				
                if (submarine.HullVertices == null) { continue; }
                //change: display the actual walls of own sub//nope, it messes up the passive
                /*if (!_.DetectSubmarineWalls)
                {
                    if (_.connectedSubs.Contains(submarine)) { continue; }                    
                }*/
                
                if ((_.CurrentMode == Sonar.Mode.Passive && submarine == Submarine.MainSub) ||(_.CurrentMode == Sonar.Mode.Active && submarine == Submarine.MainSub && _.useDirectionalPing)){ continue; }//change: don't blip the main sub in passive mode.  current workaround is to disable the blip when using directional ping
                
                //display the actual walls if the ping source is inside the sub (but not inside a hull, that's handled above)
                //only relevant in the end levels or maybe custom subs with some kind of non-hulled parts
                Rectangle worldBorders = submarine.GetDockedBorders();
                worldBorders.Location += submarine.WorldPosition.ToPoint();
                if (Submarine.RectContains(worldBorders, pingSource))
                {
                    _.CreateBlipsForSubmarineWalls(submarine, pingSource, transducerPos, pingRadius, prevPingRadius, range, passive);
                    continue;
                }
                
                //this block is for displaying submarine in both sonar in the original game
                for (int i = 0; i < submarine.HullVertices.Count; i++)
                {
                    Vector2 start = FarseerPhysics.ConvertUnits.ToDisplayUnits(submarine.HullVertices[i]);
                    Vector2 end = FarseerPhysics.ConvertUnits.ToDisplayUnits(submarine.HullVertices[(i + 1) % submarine.HullVertices.Count]);

                    if (_.item.Submarine == submarine)
                    {
                        start += Rand.Vector(500.0f);
                        end += Rand.Vector(500.0f);
                    }

                    _.CreateBlipsForLine(
                        start + submarine.WorldPosition,
                        end + submarine.WorldPosition,
                        pingSource, transducerPos,
                        pingRadius, prevPingRadius,
                        200.0f, 2.0f, range, 1.0f, passive, 
                        needsToBeInSector: needsToBeInSector);
                    }
                
            
			}

            if (Level.Loaded != null && _.item.CurrentHull != null)//remove the !detecsubwall' so it can function inside sub
            {
                if (Level.Loaded.Size.Y - pingSource.Y < range)
                {
                    _.CreateBlipsForLine(
                        new Vector2(pingSource.X - range, Level.Loaded.Size.Y),
                        new Vector2(pingSource.X + range, Level.Loaded.Size.Y),
                        pingSource, transducerPos,
                        pingRadius, prevPingRadius,
                        250.0f, 150.0f, range, pingStrength, passive, 
                        needsToBeInSector: needsToBeInSector);
                }
                if (pingSource.Y - Level.Loaded.BottomPos < range)
                {
                    _.CreateBlipsForLine(
                        new Vector2(pingSource.X - range, Level.Loaded.BottomPos),
                        new Vector2(pingSource.X + range, Level.Loaded.BottomPos),
                        pingSource, transducerPos,
                        pingRadius, prevPingRadius,
                        250.0f, 150.0f, range, pingStrength, passive, 
                        needsToBeInSector: needsToBeInSector);
                }

                List<Voronoi2.VoronoiCell> cells = Level.Loaded.GetCells(pingSource, 7);
                foreach (Voronoi2.VoronoiCell cell in cells)
                {
                    foreach (Voronoi2.GraphEdge edge in cell.Edges)
                    {
                        if (!edge.IsSolid) { continue; }
                        float cellDot = Vector2.Dot(cell.Center - pingSource, (edge.Center + cell.Translation) - cell.Center);
                        if (cellDot > 0) { continue; }

                        float facingDot = Vector2.Dot(
                            Vector2.Normalize(edge.Point1 - edge.Point2),
                            Vector2.Normalize(cell.Center - pingSource));

                        _.CreateBlipsForLine(
                            edge.Point1 + cell.Translation,
                            edge.Point2 + cell.Translation,
                            pingSource, transducerPos,
                            pingRadius, prevPingRadius,
                            350.0f, 3.0f * (Math.Abs(facingDot) + 1.0f), range, pingStrength, passive,
                            blipType : cell.IsDestructible ? Sonar.BlipType.Destructible : Sonar.BlipType.Default,
                            needsToBeInSector: needsToBeInSector);
                    }
                }
            }

            foreach (Item item in Item.SonarVisibleItems)
            {
                System.Diagnostics.Debug.Assert(_.item.Prefab.SonarSize > 0.0f);
                if (_.item.CurrentHull == null)
                {
                    float pointDist = ((_.item.WorldPosition - pingSource) * displayScale).LengthSquared();
                    if (pointDist > prevPingRadiusSqr && pointDist < pingRadiusSqr)
                    {
                        var blip = new SonarBlip(
                            _.item.WorldPosition + Rand.Vector(_.item.Prefab.SonarSize),
                            MathHelper.Clamp(_.item.Prefab.SonarSize, 0.1f, pingStrength),
                            MathHelper.Clamp(_.item.Prefab.SonarSize * 0.1f, 0.1f, 10.0f));
                        if (!IsVisible(blip)) { continue; }
                        _.sonarBlips.Add(blip);
                    }
                }
            }

            foreach (Character c in Character.CharacterList)
            {
                if (c.AnimController.CurrentHull != null || !c.Enabled) { continue; }
                if (!c.IsUnconscious && c.Params.HideInSonar) { continue; }
                if (_.DetectSubmarineWalls && c.AnimController.CurrentHull == null && _.item.CurrentHull != null) { continue; }

                if (c.AnimController.SimplePhysicsEnabled)
                {
                    float pointDist = ((c.WorldPosition - pingSource) * displayScale).LengthSquared();
                    if (pointDist > _.DisplayRadius * _.DisplayRadius) { continue; }

                    if (pointDist > prevPingRadiusSqr && pointDist < pingRadiusSqr)
                    {
                        var blip = new SonarBlip(
                            c.WorldPosition,
                            MathHelper.Clamp(c.Mass, 0.1f, pingStrength),
                            MathHelper.Clamp(c.Mass * 0.03f, 0.1f, 2.0f));
                        if (!IsVisible(blip)) { continue; }
                        _.sonarBlips.Add(blip);
                        HintManager.OnSonarSpottedCharacter(_.item, c);
                    }
                    continue;
                }

                foreach (Limb limb in c.AnimController.Limbs)
                {
                    if (!limb.body.Enabled) { continue; }

                    float pointDist = ((limb.WorldPosition - pingSource) * displayScale).LengthSquared();
                    if (limb.SimPosition == Vector2.Zero || pointDist > _.DisplayRadius * _.DisplayRadius) { continue; }

                    if (pointDist > prevPingRadiusSqr && pointDist < pingRadiusSqr)
                    {
                        var blip = new SonarBlip(
                            limb.WorldPosition + Rand.Vector(limb.Mass / 10.0f),
                            MathHelper.Clamp(limb.Mass, 0.1f, pingStrength),
                            MathHelper.Clamp(limb.Mass * 0.1f, 0.1f, 2.0f));
                        if (!IsVisible(blip)) { continue; }
                        _.sonarBlips.Add(blip);
                        HintManager.OnSonarSpottedCharacter(_.item, c);
                    }
                }
				
            }

            bool IsVisible(SonarBlip blip)
            {
                if (!passive && !_.CheckBlipVisibility(blip, transducerPos)) { return false; }
				
				if (!passive && !_.CheckBlipVisibility(blip, transducerPos)) { return false; }
				
                if (needsToBeInSector != null)
                {
                    if (!needsToBeInSector.IsWithinSector(blip.Position)) { return false; }
                }
                return true;
            }
			return false;
        }

**Comment by SomeRandomNoobKekeke** [comment]
No, directional ping is real, it doesn't trigger creatures outside the cone so i guess it's somewhere in shared
![20231216024101_2](https://github.com/Regalis11/Barotrauma/assets/122838333/91d727f1-9029-4111-8e62-56df967ec203)
I also think that the actual cone and the image on the sonar monitor may not match
https://github.com/Regalis11/Barotrauma/issues/10225

**Comment by iridescent-lightning** [comment]
Thanks. At least I know directional ping is for real. But I don't find it in shared. There is just one update method to control it on and off.

**Comment by SomeRandomNoobKekeke** [comment]
https://github.com/Regalis11/Barotrauma/blob/master/Barotrauma/BarotraumaShared/SharedSource/Items/Components/Machines/Sonar.cs#L220

https://github.com/Regalis11/Barotrauma/blob/master/Barotrauma/BarotraumaShared/SharedSource/Items/Components/Machines/Sonar.cs#L36

**Comment by iridescent-lightning** [comment]
Thanks. It has been solved. To see the results, see mod Barotrauma Die Hard.
